### PR TITLE
visitor service & persistent storage

### PR DESCRIPTION
**Pull request by oliviakim321**
- persistent storage class
- visitor service class
- make pointers to constants constant since strings are already constant
- [move imports from .h to .m files](https://stackoverflow.com/questions/2770243/do-you-put-import-s-in-h-or-m-files)
- generate or get the visitor id on sdk instantiation
- clearUser

### CODE REVIEW

File: Sources/ATTNPersistentStorage.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+- (nonnull NSString *)getPrefixedKey: (nonnull NSString * ) key;
+
+- (void)saveObject: (nonnull NSObject *) value forKey:(nonnull NSString *) key;
+
+- (nullable NSString *)readStringForKey: (nonnull NSString *) key;
+
+- (void)deleteObjectForKey: (nonnull NSString *) key;
+
+@property(nonnull, readonly) NSUserDefaults * userDefaults;
```
[the remark this answers is not in the document]
Comment (wyattjoe40) [reply Sources/ATTNPersistentStorage.h:30]: same for ATTNVisitorService

File: Sources/ATTNPersistentStorage.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+#ifndef ATTNPersistentStorage_h
+#define ATTNPersistentStorage_h
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+extern NSString * const ATTN_PERSISTENT_STORAGE_PREFIX;
+
+NS_ASSUME_NONNULL_END
```
[the remark this answers is not in the document]
Comment (wyattjoe40) [reply Sources/ATTNPersistentStorage.h:16]: same for ATTNVisitorService

File: Sources/ATTNVisitorService.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#import "ATTNPersistentStorage.h"
+
+
+#ifndef ATTNVisitorService_h
+#define ATTNVisitorService_h
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+extern NSString * const VISITOR_ID_KEY;;
```
Comment (wyattjoe40) [inline Sources/ATTNVisitorService.h:17]: does this need to be exposed in the header? I think the implementation file can just create a static const:
`static NSString *const VISITOR_ID_KEY = @"visitorId"`

File: Sources/ATTNPersistentStorage.m [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+
+NSString * const ATTN_PERSISTENT_STORAGE_PREFIX = @"com.attentive.iossdk.PERSISTENT_STORAGE";
+
+
+@implementation ATTNPersistentStorage
+
+
+- (id)init {
+    _userDefaults = [NSUserDefaults standardUserDefaults];
```
Comment (wyattjoe40) [inline Sources/ATTNPersistentStorage.m:19]: I know this is super weird, but the general pattern for an Objective-C `init` method is:

```
- (id)init {
  if (self = [super init]) {
    // any setup (like setting instance variables)
  }
  return self
}
```

It's possible the `[super init]` could return a different object instance, which is why it should be called first. 

Docs [here](https://developer.apple.com/documentation/objectivec/nsobject/1418641-init) on it.

File: Sources/ATTNVisitorService.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+@interface ATTNVisitorService : NSObject
+
+
+- (nonnull id)init;
+
+- (nonnull NSString *)getVisitorId;
+
+- (nonnull NSString *)createNewVisitorId;
+
+- (nonnull NSString *)generateVisitorId;
```
Comment (wyattjoe40) [inline Sources/ATTNVisitorService.h:31]: doesn't need to be exposed in the header

File: Sources/ATTNVisitorService.m [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+- (nonnull NSString *)createNewVisitorId {
+    NSString * newVisitorId = [self generateVisitorId];
+    [_persistentStorage saveObject:newVisitorId forKey:VISITOR_ID_KEY];
+    return newVisitorId;
+}
+
+- (nonnull NSString *)generateVisitorId {
+    NSUUID * uuid = [NSUUID UUID];
+    NSString * uuidString = [[uuid UUIDString] stringByReplacingOccurrencesOfString:@"-" withString:@""];
```
Comment (wyattjoe40) [inline Sources/ATTNVisitorService.m:40]: ez

**Review by wyattjoe40 — COMMENTED** [review]
Makes sense to me. Mostly nits, but I do think some of the public methods should be changed to private

File: Sources/ATTNPersistentStorage.m [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+//  Example
+//
+//  Created by Olivia Kim on 11/23/22.
+//
+
+#import <Foundation/Foundation.h>
+#import "ATTNPersistentStorage.h"
+
+
+NSString * const ATTN_PERSISTENT_STORAGE_PREFIX = @"com.attentive.iossdk.PERSISTENT_STORAGE";
```
[the remark this answers is not in the document]
Comment (oliviakim321) [reply Sources/ATTNPersistentStorage.m:12]: oo, I thought that `static` was for if you wanted to have separate variables for each file that imports this constant, but `extern` means that it is a global variable?

File: Sources/ATTNPersistentStorage.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+
+@interface ATTNPersistentStorage : NSObject
+
+
+- (nonnull NSString *)getPrefixedKey: (nonnull NSString * ) key;
```
[the remark this answers is not in the document]
Comment (oliviakim321) [reply Sources/ATTNPersistentStorage.h:22]: yes, good call!

File: Sources/ATTNPersistentStorage.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (wyattjoe40) [reply Sources/ATTNPersistentStorage.m:12]: I think this SO answer explains it well: https://stackoverflow.com/a/17228508

**Review by wyattjoe40 — APPROVED** [review]

**Pull request merged**